osu.Game/Online/Chat/MessageNotifier.cs

```diff
@@ -0,0 +1,234 @@
+    /// </summary>
+    public class MessageNotifier : Component
+    {
+        [Resolved(CanBeNull = true)]
+        private NotificationOverlay notificationOverlay { get; set; }
+
+        [Resolved(CanBeNull = true)]
+        private ChatOverlay chatOverlay { get; set; }
+
+        [Resolved(CanBeNull = true)]
+        private ChannelManager channelManager { get; set; }
+
+        private Bindable<bool> notifyOnMention;
+        private Bindable<bool> notifyOnChat;
+        private Bindable<User> localUser;
+
+        /// <summary>
+        /// Determines if the user is able to see incoming messages.
+        /// </summary>
+        public bool IsActive => chatOverlay?.IsPresent == true;
+
+        private readonly List<PrivateMessageNotification> privateMessageNotifications = new List<PrivateMessageNotification>();
+
+        [BackgroundDependencyLoader]
+        private void load(OsuConfigManager config, IAPIProvider api)
+        {
+            notifyOnMention = config.GetBindable<bool>(OsuSetting.ChatHighlightName);
+            notifyOnChat = config.GetBindable<bool>(OsuSetting.ChatMessageNotification);
+            localUser = api.LocalUser;
+
+            // Listen for new messages
+            channelManager.JoinedChannels.ItemsAdded += joinedChannels =>
+            {
+                foreach (var channel in joinedChannels)
+                    channel.NewMessagesArrived += channel_NewMessagesArrived;
+            };
+
+            channelManager.JoinedChannels.ItemsRemoved += leftChannels =>
+            {
+                foreach (var channel in leftChannels)
+                    channel.NewMessagesArrived -= channel_NewMessagesArrived;
+            };
+        }
+
+        private void channel_NewMessagesArrived(IEnumerable<Message> messages)
+        {
+            if (messages == null || !messages.Any())
+                return;
+
+            HandleMessages(messages.First().ChannelId, messages);
+        }
+
+        /// <remarks>
+        /// Resolves the channel id
+        /// </remarks>
+        public void HandleMessages(long channelId, IEnumerable<Message> messages)
+        {
+            var channel = channelManager.JoinedChannels.FirstOrDefault(c => c.Id == channelId);
+
+            if (channel == null)
+            {
+                Logger.Log($"Couldn't resolve channel id {channelId}", LoggingTarget.Information);
+                return;
+            }
+
+            HandleMessages(channel, messages);
+        }
+
+        public void HandleMessages(Channel channel, IEnumerable<Message> messages)
+        {
+            // don't show if the ChatOverlay and the target channel is visible.
+            if (IsActive && channelManager.CurrentChannel.Value == channel)
+                return;
+
+            foreach (var message in messages)
+            {
+                // ignore messages that already have been read
+                if (message.Id < channel.LastReadId)
+                    return;
```

Comment:
Supposed to be a `continue`, or am I wrong? Assuming we're checking messages chronologically, oldest to newest, hitting one old read message will make it so that no notifications are ever fired?

Reply:
This is intended, it doesn't make sense to check (and send notifications of) messages of a channel the user already has read.

Reply:
I understand not showing notifications for the already-read messages, but either (a) my brain is no longer working or (b) since this returns from a loop over all messages, in the scenario of
```
read message 123
read message 124
unread message 125 with a mention
```
this code will see message 123, and since it's read it will no longer check the rest, therefore not firing a notification for message 125 that it should. Unless the messages are always read *newest-first*, in which case I suppose this is OK and optimal, but warrants a comment.

Reply:
I've inversed/sorted the incoming messages by their time/ID, thanks for making me take notice of this.